### PR TITLE
[WIP] build: Align with the upstream's upgrade to Webpack 5

### PR DESCRIPTION
**Pull request by azatsarynnyy**
Signed-off-by: Artem Zatsarynnyi <azatsary@redhat.com>

### What does this PR do?
Upstream Theia has been upgraded to Webpack 5 in https://github.com/eclipse-theia/theia/pull/9451
It requires updating the format of the provided arguments: `--env arg` instead of `--env.arg`
More details are here https://github.com/eclipse/che/issues/19804#issuecomment-845377971

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/19804

### How to test this PR?

Build should pass - no error from Webpack anymore.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

File: build.include [hunk 1]
```diff
@@ -14,7 +14,7 @@ IMAGE_TAG="next"
 THEIA_GITHUB_REPO="eclipse-theia/theia"
 THEIA_VERSION="master"
 THEIA_BRANCH="master"
-THEIA_COMMIT_SHA="2eb4319fb459f30ad897e8467878154c6790aaef"
```
Comment (azatsarynnyy) [inline build.include:17]: not the latest sha, as the latest commit brings another breaking changes
I'll look at it tomorrow

**Comment by azatsarynnyy** [comment]
Still WIP. Some problems with loading the static resources
image

Continue the investigation...

File: build.include (hunk 1 above)
Comment (azatsarynnyy) [reply build.include:17]: https://github.com/eclipse/che/issues/19866

**Comment by azatsarynnyy** [comment]
[crw-ci-test --rebuild]

**Comment by dmytro-ndp** [comment]
[crw-ci-test --rebuild]

**Comment by dmytro-ndp** [comment]
[crw-ci-test --rebuild]

**Comment by azatsarynnyy** [comment]
good news - the welcome page is fixed now
bad news - I don't see the files loaded from CDN

**Comment by azatsarynnyy** [comment]
with the provided `CDN_PREFIX` and `MONACO_CDN_PREFIX`, I've got
```
[webpack-cli] TypeError: webpack.HashedModuleIdsPlugin is not a constructor
    at Object.customizeWebpackConfig (/home/theia-dev/theia-source-code/examples/assembly/cdn/webpack-customizer.js:96:33)
    at module.exports (/home/theia-dev/theia-source-code/examples/assembly/cdn/webpack.config.js:17:26)
    at /home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1391:43
    at Array.map (<anonymous>)
    at evaluateConfig (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1383:24)
    at /home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1414:53
    at async Promise.all (index 0)
    at async WebpackCLI.resolveConfig (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1413:38)
    at async WebpackCLI.createCompiler (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1837:22)
    at async WebpackCLI.buildCommand (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1952:20)
    at async Command.<anonymous> (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:742:25)
    at async Promise.all (index 1)
    at async Command.<anonymous> (/home/theia-dev/theia-source-code/node_modules/webpack-cli/lib/webpack-cli.js:1289:13)

Error: webpack exited with an unexpected code: 2.
    at ChildProcess.<anonymous> (/home/theia-dev/theia-source-code/dev-packages/application-manager/lib/application-process.js:67:28)
    at ChildProcess.emit (events.js:314:20)
    at maybeClose (internal/child_process.js:1022:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5)
error Command failed with exit code 1.
```

**Comment by azatsarynnyy** [comment]
Webpack's 4 `HashedModuleIdsPlugin` corresponds to Webpack's 5 `optimization.moduleIds: 'deterministic'`
https://webpack.js.org/migrate/5/#update-outdated-options

And `optimization.moduleIds: 'deterministic'` is default now.
https://webpack.js.org/configuration/optimization/#optimizationmoduleids

That means we don't need `HashedModuleIdsPlugin` anymore.

**Comment by azatsarynnyy** [comment]
and `html-webpack-plugin` requires updating to be compatible with Webpack 5
```
[webpack-cli] TypeError: Cannot add property htmlWebpackPluginAlterChunks, object is not extensible
    at /home/theia-dev/theia-source-code/node_modules/html-webpack-plugin/index.js:59:56
    at Hook.eval [as call] (eval at create (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:102:1)
    at Hook.CALL_DELEGATE [as _call] (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/Hook.js:14:14)
    at Compiler.newCompilation (/home/theia-dev/theia-source-code/node_modules/webpack/lib/Compiler.js:1031:26)
    at /home/theia-dev/theia-source-code/node_modules/webpack/lib/Compiler.js:1073:29
    at Hook.eval [as callAsync] (eval at create (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:4:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/Hook.js:18:14)
    at Compiler.compile (/home/theia-dev/theia-source-code/node_modules/webpack/lib/Compiler.js:1068:28)
    at /home/theia-dev/theia-source-code/node_modules/webpack/lib/Compiler.js:496:12
    at Compiler.readRecords (/home/theia-dev/theia-source-code/node_modules/webpack/lib/Compiler.js:908:11)
    at /home/theia-dev/theia-source-code/node_modules/webpack/lib/Compiler.js:493:11
    at Hook.eval [as callAsync] (eval at create (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:13:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/Hook.js:18:14)
    at /home/theia-dev/theia-source-code/node_modules/webpack/lib/Compiler.js:490:20
    at Hook.eval [as callAsync] (eval at create (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:13:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/home/theia-dev/theia-source-code/node_modules/webpack/node_modules/tapable/lib/Hook.js:18:14)

Error: webpack exited with an unexpected code: 2.
    at ChildProcess.<anonymous> (/home/theia-dev/theia-source-code/dev-packages/application-manager/lib/application-process.js:67:28)
    at ChildProcess.emit (events.js:314:20)
    at maybeClose (internal/child_process.js:1022:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5)
error Command failed with exit code 1.
```

**Comment by benoitf** [comment]
@azatsarynnyy please update as well https://github.com/eclipse-che/che-theia/blob/upstream_aligh/generator/src/templates/assembly-package.mst#L71

**Comment by azatsarynnyy** [comment]
fighting with the next problem:
```
    'ERROR in Template execution failed: ReferenceError: process is not defined\n' +
    '\n' +
    'ERROR in ReferenceError: process is not defined\n' +
    '  \n' +
    '  - custom-html.html:2946 Object.489\n' +
    '    /home/theia-dev/theia-source-code/examples/assembly/cdn/custom-html.html:294    6:15\n' +
```

**Comment by benoitf** [comment]
you may try https://github.com/SimulatedGREG/electron-vue/issues/871#issuecomment-490370237

**Comment by azatsarynnyy** [comment]
The build is fixed. Now investigating 
image

**Comment by benoitf** [comment]
ready to merge ,?

**Comment by azatsarynnyy** [comment]
> ready to merge ,?

Something went wrong with the PR status checks here, so I opened another one https://github.com/eclipse-che/che-theia/pull/1134
I want to check the build images and then will open it for a review. Today or tomorrow.

**Comment by azatsarynnyy** [comment]
superseded by https://github.com/eclipse-che/che-theia/pull/1134

**Pull request closed without merging**